### PR TITLE
Reduce some calls to re.sub

### PR DESCRIPTION
**Pull request by danielkingai2**
So calls to `re.compile` are not a problem. The main thing slowing it down is lots of calls to `re.sub` in `abbreviation_replacer.py`. I reduced some of these calls which speeds it up by a factor of ~3-3.5x on my machine, for the specific (longish) document that I tested with. I also included the script I used to test timing. Given that you are much more familiar with the codebase, see if my changes look reasonable, but all the tests do still pass. There are probably some more ways to speed up the calls in that file.

### CODE REVIEW

**Comment by nipunsadvilkar** [comment]
LGTM! Yes, I concur there are a lot of inefficient loops and calls happening in `abbreviation_replacer.py`. Since I was porting it from ruby, I thought of keeping the same loops and calls in python to get tests to pass. In a few cases, I tried to be idiomatic wherever possible, but still, there are a lot of places where we can surely improve pysbd.

It would be nice if you could list those speed up calls in the comment here so we can tackle them one by one.

File: pysbd/abbreviation_replacer.py [hunk 1]
```diff
@@ -8,20 +8,29 @@
...
 
 def replace_period_of_abbr(txt, abbr):
-    txt = re.sub(r"(?<=\s{abbr})\.(?=((\.|\:|-|\?)|(\s([a-z]|I\s|I'm|I'll|\d|\())))|(?<=^{abbr})\.(?=((\.|\:|\?)|(\s([a-z]|I\s|I'm|I'll|\d))))".format(abbr=abbr.strip()), '∯', txt)
-    txt = re.sub(r"(?<=\s{abbr})\.(?=,)|(?<=^{abbr})\.(?=,)".format(abbr=abbr.strip()), '∯', txt)
+    # prepend a space to avoid needing another regex for start of string
+    txt = " " + txt
+    txt = re.sub(r"(?<=\s{abbr})\.(?=((\.|\:|-|\?|,)|(\s([a-z]|I\s|I'm|I'll|\d|\())))".format(abbr=abbr.strip()), '∯', txt)
+    # remove the prepended space
+    txt = txt[1:]
     return txt
```
Comment (nipunsadvilkar) [inline pysbd/abbreviation_replacer.py]: Improvements for function below.
```python
def replace_abbreviation_as_sentence_boundary(txt):
    sent_starters = "|".join((r"(?=\s{}\s)".format(word) for word in SENTENCE_STARTERS))
    regex = r"(U∯S|U\.S|U∯K|E∯U|E\.U|U∯S∯A|U\.S\.A|I|i.v|I.V)∯({})".format(sent_starters)
    txt = re.sub(regex, '\\1.', txt)
    return txt
```

**Comment by danielkingai2** [comment]
I was focusing on the parts that were the majority of the time spent. Figured no need to optimize parts that weren't taking up much of the time. Here is some of the output from `cProfile`:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    1.484    1.484 <string>:1(<module>)
      366    0.003    0.000    0.258    0.001 abbreviation_replacer.py:10(replace_pre_number_abbr)
     1492    0.004    0.000    1.145    0.001 abbreviation_replacer.py:103(scan_for_replacements)
      222    0.002    0.000    0.157    0.001 abbreviation_replacer.py:19(replace_prepositive_abbr)
      904    0.007    0.000    0.726    0.001 abbreviation_replacer.py:28(replace_period_of_abbr)
        1    0.000    0.000    0.013    0.013 abbreviation_replacer.py:43(replace_abbreviation_as_sentence_boundary)
        1    0.000    0.000    0.000    0.000 abbreviation_replacer.py:54(__init__)
        1    0.000    0.000    1.400    1.400 abbreviation_replacer.py:58(replace)
        1    0.000    0.000    0.001    0.001 abbreviation_replacer.py:70(replace_multi_period_abbreviations)
        4    0.000    0.000    0.000    0.000 abbreviation_replacer.py:71(mpa_replace)
        1    0.005    0.005    1.380    1.380 abbreviation_replacer.py:83(search_for_abbreviations_in_string)
```

Didn't want to cause code churn/potentially introduce bugs by over optimizing other pieces of code. I also don't actually have clear ideas at the moment for how to make those slow pieces even faster.

File: pysbd/abbreviation_replacer.py (hunk 1 above)
Comment (danielkingai2) [reply pysbd/abbreviation_replacer.py]: Done.

**Comment by danielkingai2** [comment]
Any hesitation on releasing this as is?

**Comment by nipunsadvilkar** [comment]
Although, I wanted to make more optimizations in `abbreviation_replacer.py` but didn't find much time to do it thoroughly so releasing it for now and we can do further optimizations in the future.

Just to segregate scripts as per purpose, moving your `test_timing_script.py` to `examples` for further reference.

**Pull request merged**